Tutorial script (#3320)
ElektraInitiative/libelektra: doc/TUTORIAL_SCRIPT.md

# Script

```shell script
docker run -it elektra/elektra                                     

> # Let's map the content of a hello.js file to Elektra's configuration in /hello

Reviewer: ```suggestion
# Let's map the content of a hello.js file to Elektra's configuration
```

sudo kdb mount hello.js /hello json

# Set a value and retrieve it
kdb set user/hello/hello/world "Hello World"
kdb get user/hello/hello/world

> # You can also see where the file is located from Elektra

Reviewer: ```suggestion
# You can also see where the file is located
```

kdb file user/hello/hello/world

# You can edit the file and see what Elektra wrote
vi /home/elektra/.config/hello.js
>> Edit file to say "Hello World from File!"

# Getting the value again ... 
kdb get user/hello/hello/world

# You can also use Elektra's editor which additionally validates the file content
kdb editor user/hello ini
>> Edit file to say "Hello World from INI via kdb editor!"
kdb get user/hello/hello/world

# Elektra can export to various file formats
kdb export user/hello
kdb export user/hello yaml                                      

# You can search for plugins too
kdb plugin-list
kdb export user/hello tcl

# Let's clean up our environment before we proceed
kdb rm user/hello/hello/world

# Removing the only entry also removed the whole file
cat /home/elektra/.config/hello.ini

# Remove the mountpoint too
sudo kdb umount /hello

# Let's see how configuration validation works
# Validation works via metadata assigned to keys
# You can find all metadata by searching through METADATA.ini which is mounted like this
sudo kdb mount-info

# Now the file is mounted under system/info/elektra/metadata/#0
kdb mount

# Let's search for an integer range validation

> kdb find system/info/elektra/metadata/#0/.*/range

Reviewer: Nice!

# So check/range can be used
# Write a specification file with it
vi specification.ini                                      
>> []
>> mountpoint = specification.ini
>> [timeout]
>> check/range = 0-100

# Let's integrate the specification into the backend 
# The 'ni' plugin can be used to parse INI files into Elektra
sudo kdb mount /home/elektra/specification.ini spec/validation ni

# Let's active the configuration validation
sudo kdb spec-mount /validation

# Now let's see how it works
kdb set /validation/timeout 500

# Alright, but what if you really want a value of 500?
kdb editor spec/validation ni
> edit range to 1000

kdb set /validation/timeout 500

> 

Reviewer: Specification still not extended?

Author: I think that users will get the concept enough with one example and also when I edit the example as shown with the range editing. It is better to keep the tutorial a bit more simpler since it already is extending into a blockbuster movie length tutorial ...

# Script

```shell script
docker run -it elektra/elektra                                     

# Let's map the content of a hello.js file to Elektra's configuration
sudo kdb mount hello.js /hello json

# Set a value and retrieve it
kdb set user/hello/hello/world "Hello World"
kdb get user/hello/hello/world

# You can also see where the file is located
kdb file user/hello/hello/world

# You can edit the file and see what Elektra wrote
vi /home/elektra/.config/hello.js
>> Edit file to say "Hello World from File!"

# Getting the value again ... 
kdb get user/hello/hello/world

# You can also use Elektra's editor which additionally validates the file content
kdb editor user/hello ini
>> Edit file to say "Hello World from INI via kdb editor!"
kdb get user/hello/hello/world

# Elektra can export to various file formats
kdb export user/hello
kdb export user/hello yaml                                      

# You can search for plugins too
kdb plugin-list
kdb export user/hello tcl

> # You can see which file formats can be used to mount into Elektra

Reviewer: ```suggestion
# You can see which plugins can be used to mount configuration files into Elektra
```

# Script

```shell script
docker run -it elektra/elektra                                      > NEW LINE

> 

Reviewer: start here with an echo that mount will map the content from hello.js to Elektra's configuration in /hello.

sudo kdb mount hello.js /hello json
kdb set user/hello/hello/world "Hello World"
kdb get user/hello/hello/world
kdb file user/hello/hello/world
vi /home/elektra/.config/hello.js
# Edit file to say "Hello World from File!"
kdb get user/hello/hello/world

> kdb editor user/hello ini

Reviewer: Before doing this, write that we recommend to only use `kdb editor` as it will validate configuration?

> # Edit file to say "Hello World from INI!"

Reviewer: ```suggestion
# Edit file to say "Hello World from INI via kdb editor!"
```

kdb get user/hello/hello/world

> echo "Elektra can export to various formats!"

Reviewer: ```suggestion
echo "Elektra can export to various formats"
```

(The ! causes problems with some shells)

kdb export user/hello

> kdb export user/hello tcl

Reviewer: Use `kdb plugin-list` to find that plugin.

Reviewer: `kdb plugin-list storage` still missing

kdb export user/hello yaml                                      > NEW LINE

> echo "Remove entries and file"

Reviewer: ```suggestion
echo "Remove the only entry, which will remove the file"
```

kdb rm user/hello/hello/world
cat /home/elektra/.config/hello.ini
sudo kdb umount /hello
echo "Now let us change something.. With validation"
vi specification.ini                                       > NEW START
# content:
# []
# mountpoint = specification.ini
# [hostip]
# check/ipaddr =

> # [timeout]

Reviewer: Again: I think this is too much at once, build this up one by one (simply call spec-mount again whenever you add new meta-data to the file)

> # check/range = 0-100

Reviewer: This should be looked up in METADATA.ini (`kdb mount-info` and then `kdb meta-ls system/info/elektra/metadata/#0/check/ipaddr`)

Author: for that I would need to mount metadata.ini ... 
when i run `sudo kdb mount-info` I get no result in docker
apart from that it is a tutorial and IMO too much information anyway so it is probably better to leave this part to tutorial reading

Reviewer: yes, then put this to the tutorial. @mpranj any idea why `mount-info` cannot find METADATA.ini? #1663

Reviewer 2: I think it does work. I get the same result on my host machine and in docker:

```
elektra $ sudo kdb mount-info
elektra $ kdb meta-ls system/info/elektra/metadata/#0/check/ipaddr
comments
comments/#0
internal/ini/key/last
internal/ini/order
internal/ini/parent
internal/ini/section
```

Reviewer: Thank you for trying!

Author: Ok, this command is very confusing so I just want to add my opinion to this command which others might also think of.
1. there is no man page where I could read what it does
2. in the name is "info" which means for me that it just outputs something and does not change something (it actually mounts). A better name would be mount-defaults or something like that
3. The command does not give any output at all, not even success or what it actually has done. It should at least say it mounted the files. The "info" is especially confusing when it outputs nothing

Reviewer: I agree with your suggestions. You are in the best position to improve the situation.

Author: Sorry but I do not have time for this

sudo kdb mount /home/elektra/specification.ini spec/validation ni
sudo kdb spec-mount /validation
kdb set /validation/hostip 123.123.123.1233

> kdb set /validation/timeout 500

Reviewer: First try hostip with a valid IP again.

kdb editor spec/validation ni
#edit range to 1000
echo "Edit specification to fit needs"
kdb set /validation/timeout 500
sudo kdb umount spec/validation

> kdb rm -r /validation                                           > NEW END

Reviewer: Hosts still not done.

echo "What to do next???"
kdb
kdb --help
kdb --find-tools
```

> Not working currently

Reviewer: Why, what is the problem?
